Fixes #28217 - drop turbolinks

ezr-ondrej opened the pull request:
Drops turbolinks - takes over from #7158 (thanks @amirfefer :+1:)

ezr-ondrej committed:
Fixes #28217 - drop turbolinks

ekohl commented on docker-compose.yml:
```diff
@@ -20,7 +20,7 @@ services:
 
   app: &app_base
-    image: quay.io/foreman/foreman:develop
+    image: app/views/unattended/report_templates/ansible_inventory.erb #quay.io/foreman/foreman:develop
```
This looks incorrect

ezr-ondrej replied:
:man_facepalming: Thanks, it definitely doesn't belong here.

ekohl commented on webpack/assets/javascripts/dashboard/index.js:
```diff
@@ -24,7 +24,7 @@ function autoRefresh() {
 
       if (autoRefreshIsOn && hasFocus) {
-        window.Turbolinks.visit(window.location.href);
+        window.location.reload();
```
This will be painful I suspect. We default it to off nowadays so I suspect many users won't notice this. Probably deserves some attention.

ezr-ondrej replied:
I would guess it was painfull already. Could we just solve it without refreshing the page?

ekohl replied:
I think the dashboard is due for a proper rewrite to React and only refresh relevant content. However, that needs some thought. Most of the graphs there aren't that interesting to refresh regularly because in many use cases they don't change over the course of days. AFAIK we default to not enabling auto refresh now so it doesn't block this effort. Just good to be aware of the bigger picture.

ekohl commented on webpack/assets/javascripts/foreman_tools.js:
```diff
@@ -122,8 +122,8 @@ export function updateTable(element) {
-  /* eslint-disable no-undef */
-  Turbolinks.visit(uri.toString());
+
+  window.location.href = uri.toString();
```
Since 184f8d5ce23cfe208a696535bcea346074d8600a the entire `updateTable` function is unused in Foreman itself. Don't know about plugins but I'd appreciate it if someone would at least deprecate it but possibly just remove it.

ezr-ondrej replied:
I've deprecated it, just in case.

ekohl commented on webpack/assets/javascripts/react_app/components/SearchBar/__tests__/integration.test.js:
```diff
@@ -35,11 +36,10 @@ describe('SearchBar integration test', () => {
     const event = new KeyboardEvent('keypress', { charCode: KEYCODES.ENTER });
     global.dispatchEvent(event);
-    expect(global.Turbolinks.visit.mock.calls).toHaveLength(2);
+    // expect(global.window.location.href).not.toEqual(oldHref);
```
Not a fan of commented code. I'd prefer to drop it.

ekohl reviewed (commented):
Looks like this will affect the dashboard when autorefresh is enabled and searching (via `changeQuery`).

sharvit commented on webpack/assets/javascripts/foreman_tools.test.js:
```diff
@@ -92,9 +92,8 @@ describe('initTypeAheadSelect', () => {
-      visit: jest.fn(),
-    };
+    delete global.window.location;
+    global.window.location = { href: 'http://localhost/' };
```
Could use the `mockWindowLocation` here?

sharvit commented on webpack/assets/javascripts/foreman_tools.js:
```diff
@@ -122,12 +124,13 @@ export function updateTable(element) {
 
 // generates an absolute, needed in case of running Foreman from a subpath
 export function foremanUrl(path) {
+  /* eslint-disable-next-line no-undef */
```
`window.URL_PREFIX` instead of the comment would work?

sharvit commented on webpack/assets/javascripts/react_app/components/SearchBar/__tests__/integration.test.js:
```diff
@@ -35,11 +36,9 @@ describe('SearchBar integration test', () => {
+    expect(global.window.location.href).not.toEqual(oldHref);
     const event = new KeyboardEvent('keypress', { charCode: KEYCODES.ENTER });
     global.dispatchEvent(event);
-    expect(global.Turbolinks.visit.mock.calls).toHaveLength(2);
```
Shouldn't we expect something after pressing ENTER?

ezr-ondrej replied:
I don't know what to expect though :thinking: the url should stay the same :slightly_frowning_face:

sharvit replied:
Would it make sense spying on the `global.window.location.href` setter?
https://jestjs.io/docs/en/jest-object#jestspyonobject-methodname-accesstype

ezr-ondrej replied:
I needed a bit of codewrangling. Not sure if this is the best way, but it works :thinking: :laughing:

sharvit replied:
LGTM :+1:

sharvit reviewed (changes requested):
Thanks @ezr-ondrej

ohadlevy commented:
does it make sense to have a tubolink.visit function that is deprecated (in case some other plugin is using it?)

ezr-ondrej commented:
> does it make sense to have a tubolink.visit function that is deprecated (in case some other plugin is using it?)

Good point, it surely does make sense.

tbrisker commented on app/assets/javascripts/application.js:
```diff
@@ -20,9 +18,6 @@ $(document).on('page:fetch', function() {
 $(document).on('page:change', tfm.tools.hideSpinner);
```
This (and the above) is a turbolinks event - we should also remove anything that hooks into these events. The full list is at: https://github.com/turbolinks/turbolinks-classic#events

ezr-ondrej replied:
I've done it for core, as there was no funcionality hook on it (only cleanups), it was easy.

tbrisker commented on webpack/assets/javascripts/foreman_tools.js:
```diff
@@ -102,6 +103,7 @@ export function initTypeAheadSelect(input) {
 
 // handle table updates via turoblinks
 export function updateTable(element) {
+  deprecate('updateTable', 'react Table component', '1.26');
```
```suggestion
  deprecate('updateTable', 'react Table component', '2.1');
```

tbrisker commented on webpack/assets/javascripts/foreman_globals.js:
```diff
@@ -0,0 +1,12 @@
+    deprecate(
+      'Turbolinks.visit',
+      'react components or window.location = <url>',
+      '1.26'
```
```suggestion
      '2.1'
```

tbrisker reviewed (changes requested):
a few inline comments, otherwise good effort!

sharvit commented on webpack/assets/javascripts/foreman_globals.js:
```diff
@@ -0,0 +1,12 @@
+import { deprecate } from './react_app/common/DeprecationService';
+
+window.Turbolinks = {
```
Maybe it would make more sense to put it in `foreman_navigation`?

sharvit commented on webpack/assets/javascripts/react_app/common/MountingService.js:
```diff
@@ -6,18 +6,6 @@ export { default as registerReducer } from '../redux/reducers/registerReducer';
 
 const mountedNodes = [];
```
Should we keep managing this list now?
The whole point of this list was to be able to `unmount` all the `mountedNodes` before turbolinks mount them again.

sharvit commented on webpack/assets/javascripts/foreman_navigation.js:
```diff
@@ -1,6 +1,18 @@
 import * as LayoutActions from './react_app/components/Layout/LayoutActions';
+import { deprecate } from './react_app/common/DeprecationService';
+
+window.Turbolinks = {
```
Sorry for noticing it just now but moving it to the `foreman_navigation.js` make me wonder.
Should we add something like?
```js
export const visit = url => {
 window.location.href = url;
};

// from legacy js
window.tfm.nav.visit('some-url');

// from webpack
import { visit } from '../foreman_navigation';

visit('some-url');
```

It might help in a future refactoring.

ezr-ondrej replied:
Oh, right, thats nice! I do not really like it comming after I did it everywhere, but hey good idea, let's do it!

ezr-ondrej replied:
Done that, it wasn't that bad :) and looks much better :wink:

ekohl commented on webpack/assets/javascripts/dashboard/index.js:
```diff
@@ -2,6 +2,7 @@ import $ from 'jquery';
 import { doesDocumentHasFocus } from '../react_app/common/document';
 import { notify } from '../foreman_toast_notifications';
 import { activateTooltips, foremanUrl } from '../foreman_tools';
+import { refreshPage } from '../foreman_navigation';
```
I think you meant this:
```suggestion
import { reloadPage } from '../foreman_navigation';
```

sharvit commented:
Are the failing tests relevant?

ohadlevy commented on bundler.d/assets.rb:
```diff
@@ -1,9 +1,10 @@
   gem 'uglifier', '>= 1.0.3'
   gem 'sass-rails', '>= 5.0', ((RUBY_VERSION < '2.4') ? '< 5.0.8' : '< 6.0')
+  # this one is a dependecy for x-editable-rails
+  gem 'coffee-rails', '~> 5.0.0'
```
would be really nice to get rid of the x-editable-rails (yes i know, not in this pr :))

ezr-ondrej replied:
Yeah, that would be nice, but do we have a replacement?

ezr-ondrej commented:
> Are the failing tests relevant?

I believe not. Travis says just `eslint` missing. I don't believe that's relevant. And I think in jenkins just failed to start the server for integration-tests, but that could be related. let's see.

ezr-ondrej commented:
[test foreman]

ohadlevy commented:
I believe its a pretty simple react component? doesn't patternfly has in
place editing?

On Tue, Dec 17, 2019 at 3:23 PM Ondřej Ezr <notifications@github.com> wrote:

> *@ezr-ondrej* commented on this pull request.
> ------------------------------
>
> In bundler.d/assets.rb
> <https://github.com/theforeman/foreman/pull/7236#discussion_r358786285>:
>
> >    gem 'jquery-ui-rails', '< 5.0.0'
>    gem 'patternfly-sass', '>= 3.32.1', '< 3.38.0'
>    gem 'gettext_i18n_rails_js', '~> 1.0'
>    gem 'execjs', '>= 1.4.0', '< 3.0'
>    gem 'uglifier', '>= 1.0.3'
>    gem 'sass-rails', '>= 5.0', ((RUBY_VERSION < '2.4') ? '< 5.0.8' : '< 6.0')
> +  # this one is a dependecy for x-editable-rails
> +  gem 'coffee-rails', '~> 5.0.0'
>
> Yeah, that would be nice, but do we have an replacement?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/theforeman/foreman/pull/7236?email_source=notifications&email_token=AAAE5R2KE5MDM7ENVYQ7O6TQZDHEPA5CNFSM4JYVCIF2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCPOULVA#discussion_r358786285>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAE5R5A23PW6E4JI4IKGLLQZDHEPANCNFSM4JYVCIFQ>
> .
>

ezr-ondrej commented:
> I believe its a pretty simple react component? doesn't patternfly has in place editing?

I can't find it in their storybook, but they have it in the patternfly itself

sharvit commented:
> I can't find it in their storybook, but they have it in the patternfly itself

PF3 doesn't have it, PF4 have it in tables https://patternfly-react.surge.sh/patternfly-4/documentation/react/inline-table/inlineedit

tbrisker commented:
pushed rebase to make sure tests are green

tbrisker commented on webpack/assets/javascripts/react_app/routes/index.js:
```diff
@@ -47,7 +36,7 @@ const AppSwitcher = () => {
     const nextPath = getURIPath();
     if (currentPath !== nextPath) {
```
This is now causing an endless loop, since `currentPath` is reset to null on page load

tbrisker replied:
pushed a commit that should fix this

tbrisker committed:
Refs #28217 - correct page load without turbolinks

tbrisker commented:
travis failure is on node10 eslint and should be fixed by https://github.com/theforeman/foreman/pull/7225

tbrisker reviewed (dismissed):
Fine by me but would be good to get another ack at least on the additional commit i pushed before merging.

tbrisker commented on bundler.d/assets.rb:
```diff
@@ -1,9 +1,10 @@
   gem 'uglifier', '>= 1.0.3'
   gem 'sass-rails', '>= 5.0', ((RUBY_VERSION < '2.4') ? '< 5.0.8' : '< 6.0')
+  # this one is a dependecy for x-editable-rails
+  gem 'coffee-rails', '~> 5.0.0'
```
We need to package this gem as well until we can get rid of it

mmoll replied:
At the moment, 4.2.2 is packaged for RPM. Can we use this for now and update after https://github.com/theforeman/foreman-packaging/pull/4517 got in?

tbrisker replied:
we'll still need 5.0 for rails 6 compat

mmoll replied:
That why I said "for now" ;)

tbrisker replied:
pushed another commit with both versions dependent on rails version

tbrisker committed:
Refs #28217 - Use packaged coffee-rails gem

We currently have 4.2.2 in the rails 5.2 scl, but will need to upgrade
to 5.0 for Rails 6 compat. Using the case statement so we can start
testing Rails 6 before packaging is ready.

sharvit commented on webpack/assets/javascripts/react_app/common/testHelpers.js:
```diff
@@ -28,6 +28,20 @@ export default {
   },
 };
 
+export const mockWindowLocation = ({ href }) => {
```
Does it still need after we added `visit` to `foreman_navigation`?
looks like the `visit` method get mocked now instead and it looks better.

sharvit commented on webpack/assets/javascripts/react_app/common/urlHelpers.test.js:
```diff
@@ -44,23 +35,27 @@ describe('urlWithSearch', () => {
   const href = `${baseHref}?${oldQuery}`;
 
+  beforeEach(() => {
+    mockWindowLocation({ href });
```
Does it still need it here? Looks like the `visit` is the mock now

ezr-ondrej replied:
Well the urlHelpers work with the `window.location.href` directly, so it is necessary here.

sharvit commented on webpack/assets/javascripts/react_app/components/Layout/Layout.js:
```diff
@@ -85,9 +85,7 @@ class Layout extends React.Component {
           onItemClick={primary =>
             handleMenuClick(primary, activeMenu, changeActiveMenu)
           }
-          onNavigate={({ href, preventHref }) =>
```
The `preventHref` was here so we can use turbolinks when needed, looks like it can be removed from more places now.

ezr-ondrej replied:
Could you point me to an usage of this method? I can't see where we use it, or you mean the `preventHref` is used on different places as well? If so, could you point me to those? :) I found it only in `LayoutSelectors.js` something I've missed?

sharvit replied:
1. The `LayoutSelectors.js` uses `item.turbolinks` received from the server:
https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/components/Layout/LayoutSelectors.js#L33
Not sure if this PR should remove it, if so, there are also the `data-turbolinks-track` we use in many places.
There is a menu-items registry where we define menu items to use turbolinks :-1: 
https://github.com/theforeman/foreman/blob/develop/app/registries/menu/item.rb#L21
Looks like `turbolinks` is default true, we just use `turbolinks` for everything.

2. Looks like the `preventHref` still needed so it will use the `onNavigate` for the navigation.
See: https://github.com/patternfly/patternfly-react/blob/dbff0f181a2bd6fd763fb6769e8f9ef49281178d/packages/patternfly-3/patternfly-react/src/components/VerticalNav/VerticalNavItemHelper.js

tbrisker replied:
@ezr-ondrej looks like this comment is still not addressed? or am i missing something?

ezr-ondrej replied:
oh, I've missed this one

ezr-ondrej replied:
@sharvit we don't need to use the `preventHref` right?

sharvit replied:
Shouldn't we set it to true?
If not the browser will handle it natively? see this code from inside pf:
```js
if (preventHref && !!onClick) {
  event.preventDefault();
}
```
https://github.com/patternfly/patternfly-react/blob/dbff0f181a2bd6fd763fb6769e8f9ef49281178d/packages/patternfly-3/patternfly-react/src/components/VerticalNav/VerticalNavItemHelper.js#L88-L90

ezr-ondrej replied:
I thought we want that for pages not handled by react router, right? :thinking:

sharvit replied:
**I'm confused already but I think we should set it to `true`.**

We need the `preventDefault` when clicking a link so instead of using the native browser `href` mechanism, it would use the react-router.

If you would set it to `false`, wouldn't it use react-router while also using the native browser link?

tbrisker replied:
so what you are suggesting is we don't actually need to pass it in to the `onNavigate` function (since it isn't using it anyways) but rather explicitly add another prop below that does
```jsx
preventDefault=true
```
instead of relying on `{...this.props}` to include it and pass it on?

sharvit replied:
Yes!

ezr-ondrej commented:
I've rebased it again answered @sharvit's questions. Some other ideas?

Thanks @tbrisker for shepherding this PR :heart:

ezr-ondrej committed:
Refs #28217 - remove turbolink tags from links

tbrisker commented on app/registries/menu/item.rb:
```diff
@@ -18,7 +18,7 @@ def initialize(name, options)
       @last = options[:last] || false
       @context =  options[:engine] || Rails.application
-      @turbolinks = options.fetch(:turbolinks, true)
+      @turbolinks = false
```
do we even want to keep this option? doesn't seem like we should given it's not used. iirc this was originally introduced to allow skipping turbolinks on pages we know will cause a full page reload when loaded (due to js etc) but i'm not sure if it has ever been actually used.

ezr-ondrej replied:
dropped the option. You're right, we do not want to keep it.

tbrisker commented on webpack/assets/javascripts/react_app/components/Layout/Layout.js:
```diff
@@ -85,9 +85,7 @@ class Layout extends React.Component {
-          onNavigate={({ href, preventHref }) =>
-            history.push(href, { useTurbolinks: preventHref })
-          }
+          onNavigate={({ href }) => history.push(href)}
```
do we actually need to use `onNavigate` that basically does the same as just letting the href do it's thing?

sharvit replied:
Isn't it here so it will use the react-router `history` object?

tbrisker replied:
good point, i was confusing it with the `window.history`

tbrisker commented on webpack/assets/javascripts/react_app/components/Layout/components/__snapshots__/NavDropdown.test.js.snap:
```diff
@@ -1,7 +1,7 @@
 
 exports[`NavDropdown rendering render NavDropdown 1`] = `
-<ForwardRef
+<Uncontrolled(Dropdown)
```
is this intentional?

tbrisker replied:
guess not since the tests are failing on this

ezr-ondrej replied:
no, not really ... just blind `-u` :(

mmoll commented:
foreman test failures related :broken_heart:

ezr-ondrej commented:
[test katello]

ares commented:
tests are green now

tbrisker reviewed (changes requested):
Tests, this works fine except on the audits page - when changing page in the pagination there, an endless loop is triggered

tbrisker commented:
disregard, looks like the issue is actually in develop and I didn't switch branches before testing

tbrisker reviewed (dismissed):
Thanks @ezr-ondrej ! this works as expected, navigation between pages works correctly. Please drop turbolinks gem from packaging.

ekohl commented:
I have a script to update dependencies that I should really upstream somewhere: https://github.com/ekohl/theforeman-plumbing/blob/master/fm-packaging-update-deps. I'd recommend to merge the PR and then call the script.

ezr-ondrej commented:
> I have a script to update dependencies that I should really upstream somewhere: https://github.com/ekohl/theforeman-plumbing/blob/master/fm-packaging-update-deps. I'd recommend to merge the PR and then call the script.

I don't see any dropping of deps in the script. Maybe I just don't see it, but let's drop it manually. It is just deleting files, right?

ekohl commented:
> I don't see any dropping of deps in the script. Maybe I just don't see it, but let's drop it manually. It is just deleting files, right?

First you should update foreman.spec to drop the dependency on those packages. Since the script takes the dependencies in this repository and converts them, it should automatically drop them if they're dropped in the core repository. This PR updates `Gemfile` and `bundler.d/assets.rb` and the script will pick that up.

ezr-ondrej commented:
Packaging dropping the dependency on turbolinks: theforeman/foreman-packaging#4576

tbrisker reviewed (commented):
Tested, works well and looks like the preventHref is indeed preventing browser routing between react pages. 
Thanks @ezr-ondrej and all reviewers! Will merge once tests are green.

ezr-ondrej commented:
[test foreman][test katello] jenkins failure, reseting the tests

tbrisker commented:
[test katello]

sharvit reviewed (approved):
Thanks @ezr-ondrej, looks good and work as expected :+1:

tbrisker commented:
SQLite tests failed to execute for some reason, but this doesn't actually touch anything db-related so i'll merge.
